app/lua/lflash.c

```diff
@@ -0,0 +1,224 @@
+      NODE_DBG("%5d %5d %08x %08x %5d %1s %s\n",  
+               i, j, (size_t) ts, ts->tsv.hash, ts->tsv.len,
+               ts_isreadonly(ts) ? "R" : " ",  getstr(ts));
+    }
+} 
+
+LUA_API void dumpStrings(lua_State *L) {
+  dumpStrt(&G(L)->strt, "RAM");
+  if (G(L)->ROstrt.hash)
+    dumpStrt(&G(L)->ROstrt, "ROM");
+}
+#endif
+
+/* =====================================================================================
+ * The next 4 functions: flashPosition, flashSetPosition, flashBlock and flashErase
+ * wrap writing to flash. The last two are platform dependent.  Also note that any
+ * writes are suppressed if the global writeToFlash is false.  This is used in 
+ * phase I where the pass is used to size the structures in flash.
+ */
+static char *flashPosition(void){
+  return  flashAddr + curOffset;
+}
+
+
+static char *flashSetPosition(uint32_t offset){
+  NODE_DBG("flashSetPosition(%04x)\n", offset);
+  curOffset = offset;
+  return flashPosition();
+}
+
+
+static char *flashBlock(const void* b, size_t size)  {
+  void *cur = flashPosition();
+  NODE_DBG("flashBlock((%04x),%08x,%04x)\n", curOffset,b,size);
+  lua_assert(ALIGN_BITS(b) == 0 && ALIGN_BITS(size) == 0);
+  platform_flash_write(b, flashAddrPhys+curOffset, size);
+  curOffset += size;
+  return cur;
+}
+
+static void flashErase(uint32_t start, uint32_t end){
+  int i;
+  if (start == -1) start = FLASH_PAGES - 1;
+  if (end == -1) end = FLASH_PAGES - 1;
+  NODE_DBG("flashErase(%04x,%04x)\n", flashSector+start, flashSector+end);
+  for (i = start; i<=end; i++)
+    platform_flash_erase_sector( flashSector + i );
+}
+
+
+/* =====================================================================================
+ * Hook in user_main.c to allocate flash memory for the lua flash store
+ */
+void luaN_user_init(void) {
+  curOffset    = 0;
+  flashSector = platform_flash_reserve_section( FLASH_SIZE, &flashAddrPhys );
+  flashAddr   = cast(char *,platform_flash_phys2mapped(flashAddrPhys));
+  NODE_DBG("Flash initialised: %x %08x\n", flashSector, flashAddr); 
+}
+
+
+/*
+ * Hook in lstate.c:f_luaopen() to set up ROstrt and ROpvmain if needed
+ */  
+LUAI_FUNC void luaN_init (lua_State *L) {
+  FlashHeader *fh = cast(FlashHeader *, flashAddr);
+  if (fh->flash_sig == FLASH_SIG && 
+      fh->pROhash   != ALL_SET &&
+      fh->mainProto != ALL_SET) {
+    G(L)->ROstrt.hash = cast(GCObject **, fh->pROhash);
+    G(L)->ROstrt.nuse = fh->nROuse ;
+    G(L)->ROstrt.size = fh->nROsize;
+    G(L)->ROpvmain    = cast(Proto *,fh->mainProto);
+  }
+}
+
+#define BYTE_OFFSET(t,f) cast(size_t, &(cast(t *, NULL)->f))
+/*
+ * Rehook address chain to correct Flash byte addressed within the mapped adress space 
+ * Note that on input each 32-bit address field is split into 2×16-bit subfields
+ *  -  the lu_int16 offset of the target address being referenced
+ *  -  the lu_int16 offset of the next address pointer. 
+ */
+
+static int rebuild_core (int fd, uint32_t size, lu_int32 *buf) {
```

Comment:
I'd love for this to be more easily exposed (so not `static` for starters). The way we'd like to use the LFS feature at $work is to pre-build the entire image and flash (or OTA-upgrade) that, with the actual Lua code not even knowing how to `node.flash.reload()`.

I'd originally envisaged just adding a flag to `luac.cross` to spit out non-PIC addresses, but having seen the format now, I'm thinking it would be easier to write a separate utility that post-processes the `flash.img` from PIC to relocated-ready-to-be-dumped-into-flash format. That will need `rebuild_core` though (possibly with `flashBlock` passed in as a function pointer for flexibility).

Reply:
I think that I mentioned somewhere that `luac.cross` is so fast that it easier just to add a `-a` switch to fix the LFS at a fixed address, and run the `luac.cross` multiple times if necessary.

This still leaves the need to do address calculation in your make.  The alternative is to embed this functionality in a smarter esptool.

Reply:
As far as provisioning goes, I would class DiUS and me as advanced users, each with our own sweet spot.

In my case I have my own provisioning system.  I _very_ rarely want to change the base firmware (for example my home automation system is still running a build of dev from last Aug).  But I do use my own provisioning system a _lot_.  This probes a flag byte in RTCmem and if set then the ESP end does a reprovision on reboot.  I've integrated this into my ESP apps as a reprovision command, so each ESP config has its own master directory hierarchy on the host provisioning server.  If I want to make a change then I just change the master file on the server and issue a preprovision command to the ESP app, and bang -- magic happens -- and a few seconds later the ESP is rebooted with the new app running.  I am still testing the LFS changes but this is all pretty transparent to the app -- other than LFS aallows you to code your apps in a simpler cleaner way because you don't have to worry about the code footprint in RAM.

However, all this does rely on you having a network path from each ESP to some provisioning server.  So this isn't a good solution for a lot of embedded IoT uses.

The way that I currently do the bootstrap is to have a `init.lua` stub which either runs the `node.flash.reload()` if LFS isn't loaded or it chains into the init module in the LFS.  From there on in, all encapsulation is done by Lua features.

Reply:
If you think a `-a` switch is the better approach, sure, I don't mind.